include/flatbuffers/flatbuffers.h

```diff
@@ -542,49 +547,181 @@ struct String : public Vector<char> {
+// BufferDeleter is specific to the `release` functionality, and allows the
+// user to safely gain ownership of a buffer and its associated buffer.
+class BufferDeleter {
+ public:
+  BufferDeleter(std::unique_ptr<Allocator> allocator, uint8_t *buf, size_t size)
+      : allocator_(std::forward<std::unique_ptr<Allocator>>(allocator)),
+        buf_(buf),
+        size_(size) {
+    assert(allocator_);
+  }
+
+  BufferDeleter(BufferDeleter &&other)
+      : allocator_(std::forward<std::unique_ptr<Allocator>>(other.allocator_)),
+        buf_(other.buf_),
+        size_(other.size_) {
+    assert(allocator_);
+  }
+
+  inline void operator()(uint8_t * /* unused */) {
+    allocator_->deallocate(buf_, size_);
+  }
+
+  // must stay at end of public section, may change access to private
+  FLATBUFFERS_DELETE_FUNC(BufferDeleter(const BufferDeleter& other))
+  FLATBUFFERS_DELETE_FUNC(BufferDeleter& operator=(const BufferDeleter& other))
+
+ protected:
+  std::unique_ptr<Allocator> allocator_;
+  uint8_t *buf_;
+  size_t size_;
+};
+
+// Pointer to relinquished memory.
+#if (!defined(_MSC_VER) || _MSC_VER > 1600)
+typedef std::unique_ptr<uint8_t, BufferDeleter> unique_ptr_t;
+#else  // begin hack for MSVC 16
+class unique_ptr_t {
+ public:
+  inline unique_ptr_t(uint8_t *ptr, BufferDeleter &&deleter)
+    : ptr_(ptr), deleter_(std::forward<BufferDeleter>(deleter)) {
+  }
+
+  inline unique_ptr_t(unique_ptr_t &&other)
+      : ptr_(other.ptr_), deleter_(std::forward<BufferDeleter>(other.deleter_)) {
+  }
+
+  inline ~unique_ptr_t() {
+    if (ptr_ != nullptr) {
+      deleter_(ptr_);
+    }
+  }
+
+  inline uint8_t *release() FLATBUFFERS_NOEXCEPT {
+    if (ptr_ != nullptr) {
+      deleter_(ptr_);
+    }
+    ptr_ = nullptr;
+    return nullptr;
+  }
+
+  inline void reset(std::nullptr_t /* unused */ = nullptr) FLATBUFFERS_NOEXCEPT {
+    if (ptr_ != nullptr) {
+      deleter_(ptr_);
+    }
+    ptr_ = nullptr;
+  }
+
+  inline uint8_t *get() const FLATBUFFERS_NOEXCEPT {
+    return ptr_;
+  }
+
+  inline uint8_t operator*() const {
+    return *ptr_;
+  }
+
+  inline uint8_t *operator->() const FLATBUFFERS_NOEXCEPT {
+    return ptr_;
+  }
+
+ protected:
+  unique_ptr_t(const unique_ptr_t& other);  // delete
+  unique_ptr_t& operator=(const unique_ptr_t& other);  // delete
+
+  uint8_t *ptr_;
+  BufferDeleter deleter_;
+};
+#endif  // end hack for MSVC 16
+#endif  // FLATBUFFERS_CPP98_STL
+
 // This is a minimal replication of std::vector<uint8_t> functionality,
 // except growing from higher to lower addresses. i.e push_back() inserts data
 // in the lowest address in the vector.
 class vector_downward {
  public:
   explicit vector_downward(size_t initial_size,
-                           const simple_allocator &allocator)
-    : reserved_((initial_size + sizeof(largest_scalar_t) - 1) &
+                           std::unique_ptr<Allocator> &&allocator)
+
+    : allocator_(std::forward<std::unique_ptr<Allocator>>(allocator)),
```

Comment:
I think that's a bit much for something that's for very few users and dying out. A typedef to have either naked or unique would be sufficient I think.